File: 프로그래머스/2/42860. 조이스틱/조이스틱.java
import java.util.*;

class Solution {
    public int solution(String name) {
        int answer = 0;
        
        
        // 위아래 알파벳 최소 변경
        for(char c : name.toCharArray()) {
            int result = alphaMove(c);
            answer+= result;
        }
        
        
        // 왼오 커서 위치 최소 변경
        // 1 ~ 10
        // 1          7 8 9
        // 1 -> 9 (2) + 9-> 8 + 8 -> 7 => 4
        // 1 -> 7 (6) + 7 -> 8 + 8 -> 9 => 8
        ArrayList<Integer> list = new ArrayList<>();
        for(int i = 0; i < name.length(); i++) {
            if(name.charAt(i) != 'A'){ // 'A'가 아니면
                list.add(i); // index값 list에 추가
            }
        }
        
        int count = positionMove(name);
        answer += count;
        
        
        return answer;
    }
    
    public int positionMove(String name) {
        // list = 1 7 8 9 
        int n = name.length();
        int move = n -1;
        
        for(int i = 0; i < n; i++) {
            int next = i +1;
            
            while(next < n && name.charAt(next)=='A') {
                next++;
            }
            
            // 오른쪽으로 갔다가 되돌아오기
            int case1 = i * 2 + (n - next);
            
            // 왼쪽 먼저 갔다가 오기
            int case2 = i + 2 * (n-next);
            
            move = Math.min(move, Math.min(case1, case2));
        }
        
        return move;
    }
    
    public int alphaMove(char c) {
        int A = 'A' - '0';
        int Z = 'Z' - '0';
        int C = c - '0';
        
        return Math.min(C-A, Z-C+1); // (오른쪽 , 왼쪽)
    }
}

File: 프로그래머스/2/42860. 조이스틱/조이스틱.py
def solution(name):
    answer = 0
    pos = []
    n = len(name)
    
    # 위/아래 조작 개수 세기
    for ch in name:
        answer += min(ord(ch) - ord('A'), ord('Z') - ord(ch) + 1)

        
    # 오/왼 조작 개수 세기
    move = n -1
    for i in range(n):
        next_i = i+1
        # A가 연속되는 구간 찾기
        while next_i < n and name[next_i] == 'A':
            next_i += 1
            
        # 좌로 갔다가 우로 돌아오거나 (i + i + n - next_i)
        # 우로 갔다가 좌로 돌아오는 경우
        move = min(move, 2 * i + n - next_i)
        move = min(move, i + 2 * (n-next_i))
        
    return answer + move